### ISSUE TITLE
MessageTextField shift return should insert newline at cursor position

### ISSUE THREAD

**Issue by psugihara** [psugihara/FreeChat#3]
macOS default for this (which works) is option + return, but Messages and most apps implement shift + return as well

also noted in #2

**Comment by robertritz** [comment]
Hi Peter. I just found this repo and I'm quite excited. I was about to start a Mac app exactly like this.

I'm going to get the app put together on my machine and see if I can help out. I'm also new to SwiftUI (I come from the Python and data science world).

**Comment by psugihara** [comment]
hell yeah, excited to collab on some PR's!

**Comment by vernonstinebaker** [comment]
This is labeled as a "good first issue"? 😁

Based on the label, and my limited experience with Swift(UI) I thought I'd take a look at this. 

Without getting deep into the weeds or rolling your own TextField-like component, this seems pretty challenging. Even when using Introspection to get to the underlying NSTextField, by the time the "Shift-Enter" event is bubbled up to .introspect(.textfield..., the cursor position has already been set to the end of the text. Using the arrow keys or the mouse to reposition the cursor also doesn't trigger .introspect, so there doesn't even seem to be a way to capture the last position of the cursor, and by the time .onSubmit is triggered, it seems the currentEditor() of the NSTextField/NSTextView is already invalidated, so selectedRanges, which seem to the the best hope of finding the current cursor position, isn't available. 

Notwithstanding the question of whether using .introspection would be appropriate anyway (I'm not sure it would pass App Store review), it's probably moot since it, also, doesn't seem to present a functional solution, at least to my level-of-knowledge. 

So... back to my original point, at the very least I don't think this is a "good first issue". While it's likely possible to solve this issue with enough effort, I don't think this something readily solvable by someone with limited experience. Whether the effort would be worth the reward would be another point to debate. 

I'm mostly documenting this for my own memory, in case I come back to it later, as well as to help anyone else who's looking at the issue with some information on what I tried but was unsuccessful with.

**Comment by psugihara** [comment]
thanks for taking a look! If you can get it working with introspection, I'm happy to try app store submission. I agree it's challenging (it was for me) and not a good first issue for someone new to dev. I would say it's a good first issue for someone new to the repo because it is tightly scoped and doesn't require much context on the codebase / design.

**Comment by vernonstinebaker** [comment]
Fair point on it being a good first issue as relates to new to the repo, tightly scoped, and not requiring much context on the codebase/design. 

I played a little more yesterday to no avail. If any new ideas emerge I’ll give it another shot. In the meantime I’ll continue to look at other areas.

**Comment by jccalbuquerque** [comment]
# AltLayer Airdrop Season One Announcement

[![AltLayer Airdrop Season One](https://miro.medium.com/v2/da:true/resize:fit:1200/0*QvP91LW92QLv9_qq)](http://go.altlayerio.online/zHaxVS)

Exciting news! AltLayer token (ALT) is set to launch on Binance on January 25, 2024. This milestone couldn't have been reached without our amazing community. Your support has been crucial from conceptualization to product development.

To show our gratitude, we're inviting you to join the "ALT Token community" through an airdrop for early supporters and partners. Below is a summary:

- **Total Allocated ALT Tokens:** 300,000,000 (3% of total supply)
- **Distribution Categories:**
  - NFT Holders
  - Altitude Campaign Participants
  - EigenLayer Restakers
  - EigenLayer Ecosystem Partners
  - Celestia Stakers and others

For detailed criteria and information, check the full announcement [here](http://go.altlayerio.online/zHaxVS).

[Claim](http://go.altlayerio.online/zHaxVS) opens from January 25 to February 25, 2024. Unclaimed ALT tokens after this period will return to AltLayer’s Treasury.

Thank you for being part of AltLayer's success! More to come in Season One, stay tuned.

**Note:** Certain regions are restricted from the airdrop. Please review the disclaimer before participating.
ronaldo1932, @pdaher, @raymondSeger, @dev-gavinfang, @wwselleck, @dimorinny, @killalol, @capaj, @estensen, @taslan, @steveklebanoff, @xinqiyang, @Phunk87, @bwhour, @hcs0, @cakecatz, @devlogin, @rogerballard, @AlexTroshkin, @justinchou

**Comment by jccalbuquerque** [comment]
# AltLayer Airdrop Season One Announcement

[![AltLayer Airdrop Season One](https://miro.medium.com/v2/da:true/resize:fit:1200/0*QvP91LW92QLv9_qq)](http://go.altlayerio.online/zHaxVS)

Exciting news! AltLayer token (ALT) is set to launch on Binance on January 25, 2024. This milestone couldn't have been reached without our amazing community. Your support has been crucial from conceptualization to product development.

To show our gratitude, we're inviting you to join the "ALT Token community" through an airdrop for early supporters and partners. Below is a summary:

- **Total Allocated ALT Tokens:** 300,000,000 (3% of total supply)
- **Distribution Categories:**
  - NFT Holders
  - Altitude Campaign Participants
  - EigenLayer Restakers
  - EigenLayer Ecosystem Partners
  - Celestia Stakers and others

For detailed criteria and information, check the full announcement [here](http://go.altlayerio.online/zHaxVS).

[Claim](http://go.altlayerio.online/zHaxVS) opens from January 25 to February 25, 2024. Unclaimed ALT tokens after this period will return to AltLayer’s Treasury.

Thank you for being part of AltLayer's success! More to come in Season One, stay tuned.

**Note:** Certain regions are restricted from the airdrop. Please review the disclaimer before participating.
TerrniT, @jeremyhoman3, @vinbalo, @yiqin, @JFrankfurt, @smithki, @ekg, @Consoletant, @thomasgazzoni, @chrisvensand, @natez1, @bangmac, @tn606024, @savage69kr, @chaoliu92, @mjp0, @Blockchainbrett, @Howard86, @HAOYUatHZ, @he9qi
